sky/tools/sky_snapshot/main.cc

```diff
@@ -1,108 +1,240 @@
+#include "lib/tonic/converter/dart_converter.h"
+#include "lib/tonic/file_loader/file_loader.h"
+
+extern "C" {
+extern const uint8_t* kDartVmIsolateSnapshotBuffer;
+extern const uint8_t* kDartIsolateSnapshotBuffer;
+}
+
+namespace sky_snapshot {
+namespace {
+
+using tonic::ToDart;
+
+constexpr char kHelp[] = "help";
+constexpr char kPackages[] = "packages";
+constexpr char kSnapshot[] = "snapshot";
+constexpr char kDepfile[] = "depfile";
+constexpr char kBuildOutput[] = "build-output";
+
+const char* kDartArgs[] = {
+    // clang-format off
+    "--enable_mirrors=false",
+    "--load_deferred_eagerly=true",
+    "--conditional_directives",
+    // TODO(chinmaygarde): The experimental interpreter for iOS device targets
+    // does not support all these flags. The build process uses its own version
+    // of this snapshotter. Till support for all these flags is added, make
+    // sure the snapshotter does not error out on unrecognized flags.
+    "--ignore-unrecognized-flags",
+    // clang-format on
+};
 
 void Usage() {
-  std::cerr << R"USAGE(usage: sky_snapshot --packages=PACKAGES --snapshot=SNAPSHOT_BLOB <lib/main.dart>
-       [ --depfile=DEPS_FILE ] [ --build-output=BUILD_OUTPUT ] [ --help ]
-
- * PACKAGES is the '.packages' file that defines where to find Dart packages.
- * SNAPSHOT_BLOB is the file to write the snapshot into.
- * DEPS_FILE is an optional '.d' file to depedendency information into.
- * BUILD_OUTPUT optionally overrides the target name used in the DEPS_FILE.
-   (The default is SNAPSHOT_BLOB.)
-)USAGE";
+  std::cerr
+      << "Usage: sky_snapshot --" << kPackages << "=PACKAGES" << std::endl
+      << "                  [ --" << kSnapshot << "=OUTPUT_SNAPSHOT ]"
+      << std::endl
+      << "                  [ --" << kDepfile << "=DEPFILE ]" << std::endl
+      << "                  [ --" << kBuildOutput << "=BUILD_OUTPUT ]"
+      << std::endl
+      << std::endl
+      << "                        MAIN_DART" << std::endl
+      << " * PACKAGES is the '.packages' file that defines where to find Dart "
+         "packages."
+      << std::endl
+      << " * OUTPUT_SNAPSHOT is the file to write the snapshot into."
+      << std::endl
+      << " * DEPFILE is the file into which to write the '.d' depedendency "
+         "information into."
+      << std::endl
+      << " * BUILD_OUTPUT determines the target name used in the " << std::endl
+      << "   DEPFILE. (Required if DEPFILE is provided.) " << std::endl;
+}
+
+class DartScope {
+ public:
+  DartScope(Dart_Isolate isolate) {
+    Dart_EnterIsolate(isolate);
+    Dart_EnterScope();
+  }
+
+  ~DartScope() {
+    Dart_ExitScope();
+    Dart_ExitIsolate();
+  }
+};
+
+void InitDartVM() {
+  FTL_CHECK(Dart_SetVMFlags(arraysize(kDartArgs), kDartArgs));
+  char* error = Dart_Initialize(
+      kDartVmIsolateSnapshotBuffer, nullptr, nullptr, nullptr, nullptr, nullptr,
+      nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr);
```

Comment:
i am a dart vm user, let me sing you the song of my people

null null null null null null null null null null null null